ftprime/argrecorder.py

```diff
@@ -196,6 +208,34 @@ def get_nodes(self, input_ids):
         """
         return [self.node_ids[j] for j in input_ids]
 
+    def add_individuals(self, input_ids, times, flagss=None, populations=None):
+        '''
+        Add new individuals, resulting in records as in `.add_individual`, but
+        more efficiently.
+
+        :param iterable of int input_ids: The input ID of the new individual.
+        :param iterable of float times: The time of birth of the individual.
+        :param iterable of int flagss: Any msprime flags to record.
+        :param iterable of int populations: The population ID of birth of the
+            individual.
+        '''
+        if flagss is None:
+            flagss = itertools.cycle((msprime.NODE_IS_SAMPLE, ))
+        if populations is None:
+            populations = itertools.cycle((msprime.NULL_POPULATION, ))
+
+        # inflate to a list so we have a length; then set up the map from
+        # node_ids to tree sequence samples
+        input_ids = list(self._valid_input_ids(input_ids))
```

Comment:
This could be done more efficiently in numpy, I think - first load these things into `nt` as below; then check if any exist in `self.node_ids` already?

Reply:
it may also be worth checking if substantial time is spent here.